scalding-core/src/main/scala/com/twitter/scalding/macros/impl/FieldsProviderImpl.scala

```diff
@@ -80,59 +80,104 @@ object FieldsProviderImpl {
+          case tpe if tpe.erasure =:= typeOf[Option[Any]] =>
+            val innerType = tpe.asInstanceOf[TypeRefApi].args.head
+            matchField(innerType, outerName, fieldName, true)
+          case tpe if (tpe.typeSymbol.isClass && tpe.typeSymbol.asClass.isCaseClass) =>
+            val prefix = outerName.map(pre => s"$pre$fieldName.")
+            expandMethod(tpe, prefix, isOption)
+          case tpe if allowUnknownTypes => simpleRet
+          case _ =>
+            c.abort(c.enclosingPosition, s"Case class ${T} is not pure primitives or nested case classes")
+        }
+      }
 
-    def expandMethod(outerTpe: Type, outerName: Option[String], isOption: Boolean): List[(Tree, String)] = {
-      outerTpe
-        .declarations
-        .collect { case m: MethodSymbol if m.isCaseAccessor => m }
-        .flatMap { accessorMethod =>
-          val fieldName = accessorMethod.name.toTermName.toString
-          val fieldType = accessorMethod.returnType
-          matchField(fieldType, outerName, fieldName, isOption)
-        }.toList
-    }
+      def expandMethod(outerTpe: Type, outerName: Option[String], isOption: Boolean): List[(Tree, String)] = {
+        outerTpe
+          .declarations
+          .collect { case m: MethodSymbol if m.isCaseAccessor => m }
+          .flatMap { accessorMethod =>
+            val fieldName = accessorMethod.name.toTermName.toString
+            val fieldType = accessorMethod.returnType.asSeenFrom(outerTpe, outerTpe.typeSymbol.asClass)
 
-    val prefix = if (namingScheme == NamedNoPrefix) None else Some("")
-    val expanded = expandMethod(T.tpe, prefix, false)
-    if (expanded.isEmpty) c.abort(c.enclosingPosition, s"Case class ${T} has no primitive types we were able to extract")
+            matchField(fieldType, outerName, fieldName, isOption)
+          }.toList
+      }
+
+      val prefix = if (namingScheme == NamedNoPrefix) None else Some("")
+      val expanded = expandMethod(T.tpe, prefix, false)
+      if (expanded.isEmpty) c.abort(c.enclosingPosition, s"Case class ${T} has no primitive types we were able to extract")
 
-    val typeTrees = expanded.map(_._1)
+      Left(expanded)
+    }
 
+    val typeTrees = expanded.fold({ list => list.map(_._1) }, { list => list.map(_._1) })
     val res = if (namingScheme == NamedWithPrefix || namingScheme == NamedNoPrefix) {
-      val fieldNames = expanded.map(_._2)
-      q"""
-      new _root_.cascading.tuple.Fields(_root_.scala.Array.apply[java.lang.Comparable[_]](..$fieldNames),
-        _root_.scala.Array.apply[java.lang.reflect.Type](..$typeTrees)) with _root_.com.twitter.bijection.macros.MacroGenerated
-      """
+      val leftFn = { fieldNames: List[String] =>
+        q"""
+        new _root_.cascading.tuple.Fields(_root_.scala.Array.apply[_root_.java.lang.Comparable[_]](..$fieldNames),
+          _root_.scala.Array.apply[java.lang.reflect.Type](..$typeTrees)) with _root_.com.twitter.bijection.macros.MacroGenerated
```

Comment:
I don't think I see the benefit in adding with MacroGenerated on the Fields instance, since this will cause us to create a new class for each macro, whereas we could have just created an instance. I think we should remove this.